tinyDA/proposal.py

```diff
@@ -1594,3 +1619,31 @@ def get_acceptance(
             + previous_link_below.posterior
             - proposal_link_below.posterior
         )
+
+
+class DREAM(DREAMZ, SharedArchiveProposal):
+    def __init__(
+        self,
+        M0,
+        delta=1,
+        b=5e-2,
+        b_star=1e-6,
+        Z_method="random",
+        nCR=3,
+        adaptive=False,
+        gamma=1.01,
+        period=100,
+    ):
+        DREAMZ.__init__(self, M0, delta, b, b_star, Z_method, nCR, adaptive, gamma, period)
+        SharedArchiveProposal.__init__(self)
+
+    def adapt(self, **kwargs):
```

Comment:
Maybe we could do something like
```
def adapt(self, **kwargs):
        super().adapt(**kwargs)
        self.update_archive(kwargs["parameters"])

def make_proposal(self, link):
        Z = self.read_archive()
        return super().make_proposal(link, Z)
```
and then make changes to accommodate for passing in an archive to the parent class `DREAMZ`. That would make the crossover probabilities `pCR` local to each instance of the proposal, but I don't think it would mean much in practice.

Reply:
Not sure I understand this suggestion. From what I understand, `pCR` is static if proposal is set to non-adaptive and changes when the proposal adapts if it is adaptive. I assume the issue here is every proposal would have the same variance of the archive `Z` because its the shared one.

I did change DREAM now so that every instance of the proposal has a copy of its initial `ZL` archive and when setting up the true archive `Z`, it now adds together its own initial archive `ZL` and the shared archive `ZS`. This way every proposal instance has a different variance according to its true archive `Z`. The shared archive now only contains the samples outside of the initial archives.

Reply:
@bagr-sus what I meant was that the `adapt` method of each parallel proposal wouldn't necessarily have to know about the global archive. I think it's fine if the adaptations are based only on the local history, i.e. the original `self.Z` of `DREAMZ`.

My suggestion was to only access the global archive inside `make_proposal`. That way you wouldn't have to keep a copy of the entire archive in memory for each proposal, and only have to load it during `make_proposal`. With a very small change to the parent class `DREAMZ` this could be much simpler to do.

I made a gist to better illustrate what I mean: https://gist.github.com/mikkelbue/7d88cdb2c6cea879071c89bcf07b37ad
`DREAMZ.make_proposal` now takes an optional argument which is the archive that is used to construct the proposal.

Reply:
Thanks for taking the time to write up an explanation with a code example. I understand your suggestions now.

I've basically moved the code 1:1 into the branch, the only notable added change was modifying the shared archive a bit again. Now, every instance of the proposal contributes its own initial archive into the shared archive. I'm not sure whether that changes much, other than making the initial archive bigger. 

The reason for this change was that at the start of sampling the shared archive would be empty, and therefore it wouldn't be possible to select the random samples for the jump vector. Alternatively, the proposal's initial archive could be used to compute the first (or first `n`) samples.